syntax = "proto3";
import "common.proto";

package proto.ext.application;

message OpenApplicationRequest {
  string target = 1;
  optional string app_id = 2;
};

message RunInTerminalRequest {
  repeated string cmdline = 1;
  optional string app_id = 2;
  optional string working_directory  = 3;
  optional string title = 4;
  bool hold = 5;
};

message ListApplicationRequest {
  optional string target = 1;
};

message ListApplicationResponse {
  repeated Application apps = 1;
};

message GetDefaultApplicationRequest {
  string target = 1;
};

message GetDefaultApplicationResponse {
  Application app = 1;
};

message Request {
  oneof payload {
    ListApplicationRequest list = 1;
    OpenApplicationRequest open = 2;
    GetDefaultApplicationRequest get_default = 3;
    RunInTerminalRequest run_in_terminal = 4;
  };
};

message Response {
  oneof payload {
    ListApplicationResponse list = 1;
    common.AckResponse open = 2;
    GetDefaultApplicationResponse get_default = 3;
    common.AckResponse run_in_terminal = 4;
  };
};

message Application {
  string id = 1;
  string name = 2;
  string icon = 3;
  string path = 4;
};
